# Experiment Size

In this notebook I want to illustrate the basic idea behind determining a minimum sample size for a given test.

**Case Explanation**

* You are a online retailer and want to test a new design of your landing page. 
* Your hope is that this change will increase your conversion rate from .12 to .14. 
* The probability of rejecting a true null hypothesis should be .05.
* The statistical power of the test should be .80.

What is the minimum sample size for the experiment?

**Hypothesis**

Let $p_t$ be the conversion rate in the test group and $p_c$ the conversion rate in the control group.

H0: $p_c = p_t = 0.12$

H1: $p_c = 0.12$ and $p_t = 0.14$

**Sampling distribution of p**

The distribution of the conversion rate is a special case of the sampling distribution of the mean. The total number of conversions in a given sample follows a binomial distribution with the parameters $p$ and $n$. The conversion rate gives the mean number of conversions in a given sample. And for large enough sample sizes, it follows a normal distribution with the parameters $\mu = p$ and $\sigma^2 = \frac{p*(1-p)}{n}$.

**Test statistic**

We want to test whether there is a significant difference between the conversion rates in both groups. Therefore, we construct the test statistic $d$, measuring the conversion rate difference between test and control group: 

$d = p_t - p_c$

Under H0, we assume that $p_t$ = $p_c = 0.12$. The sampling distribution of $d$ is then:

$d_0 \sim N(\mu = 0, \sigma^2 = 2*\frac{p_t*(1-p_t)}{n})$

Under H1, we assume that $p_c = 0.12$ and $p_t = 0.14$. The sampling distribution of d is then:

$d_1 \sim N(\mu = p_t - p_c, \sigma^2 = \frac{p_t*(1-p_t)+p_c*(1-p_c)}{n})$

**Experiment size**

We can calculate the minimum sample size by creating equations for $\alpha$ (Type I error) and $\beta$ (Type II error) of the test.

**Type I Error**

$P(\text{ Reject H0 } | \text{ H0 is true }) = \alpha$

$d_{min} = z_{1-\alpha} * \sqrt{\frac{2*p_t*(1-p_t)}{n}}$

**Type II Error**

$P(\text{ Accept H0 } | \text{ H1 is true }) = \beta$

$d_{min} - (p_t - p_c) = z_{\beta} * \sqrt{\frac{p_t*(1-p_t)+p_c*(1-p_c)}{n}}$


By relating the two equations we get:

$n = \lceil \big(\frac{z_{\alpha} ( 2 * p_t * (1-p_t) ) - z_{\beta} ( p_t * (1-p_t) + p_c * (1-p_c) )}{p_t - p_c}\big)^2 \rceil$

In [4]:
import scipy.stats as stats
import numpy as np


def determine_experiment_size(p_null, p_alt, alpha, beta):
    """Returns the minimal sample size necessary for a specified
    test
    
    Input:
        p_null (float): conversion rate under H0
        p_alt (float): conversion rate under H1
        alpha: probability of Type I error
        beta: probability of Type II error
    
    Returns:
        float: minimal sample size
    
    """
    
    # Get necessary z-scores
    z_null = stats.norm.ppf(1 - alpha)
    z_alt  = stats.norm.ppf(beta)

    # Get standard deviations
    sd_null = np.sqrt(p_null * (1-p_null) + p_null * (1-p_null))
    sd_alt  = np.sqrt(p_null * (1-p_null) + p_alt  * (1-p_alt) )

    # Calculate minimum sample size
    p_diff = p_alt - p_null
    n = ((z_null*sd_null - z_alt*sd_alt) / p_diff) ** 2
    
    return np.ceil(n)

In [7]:
determine_experiment_size(0.12, 0.14, 0.05, 0.2)

3341.0

For the specified test we would need at least 3341 observations per group.